src/android/FingerprintAuthenticationDialogFragment.java

```diff
@@ -78,6 +78,12 @@ public void onCreate(Bundle savedInstanceState) {
     @Override
     public View onCreateView(LayoutInflater inflater, ViewGroup container,
                              Bundle savedInstanceState) {
+        if (!Fingerprint.isPluginSetUp()) {
+            Log.i(TAG, "Invalid state, view not created");
+            dismissAllowingStateLoss();
```

Comment:
```suggestion
            Fingerprint.onCancelled();
            dismissAllowingStateLoss();
```
Since we are closing the dialog here the app/Javascript part needs to be notified too, because we could build a security hole here.

Reply:
Can you please test if this works or the app crashes? We need to understand more what happens in this case.

Reply:
The app crashes with this change. As I told you, OS clears up the application as well as the plugin state because of memory and when the user tries to resume it, OS tries to restore the application state but the app crashes if we access to some variable we think it is set up. All plugin logic attached to an Android hook lifecycle could throw an exception if this situation takes place and for that reason I decided to cut off the plugin execution as soon I detect an unexpected plugin state (package is null).

This is the exception thrown if we invoke 'Fingerprint.onCancelled' instead of 'dismissAllowingStateLoss'. The app is being restarted so none is listening to the plugin.

`04-15 08:48:51.711 32430 32430 D AndroidRuntime: Shutting down VM
04-15 08:48:51.718 32430 32430 E AndroidRuntime: FATAL EXCEPTION: main
04-15 08:48:51.718 32430 32430 E AndroidRuntime: Process: com.okode.demo, PID: 32430
04-15 08:48:51.718 32430 32430 E AndroidRuntime: java.lang.RuntimeException: Unable to start activity ComponentInfo{com.okode.demo/com.okode.demo.MainActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'void org.apache.cordova.CallbackContext.error(java.lang.String)' on a null object reference
04-15 08:48:51.718 32430 32430 E AndroidRuntime: 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2957)
04-15 08:48:51.718 32430 32430 E AndroidRuntime: 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3032)
04-15 08:48:51.718 32430 32430 E AndroidRuntime: 	at android.app.ActivityThread.-wrap11(Unknown Source:0)
04-15 08:48:51.718 32430 32430 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1696)
04-15 08:48:51.718 32430 32430 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:105)
04-15 08:48:51.718 32430 32430 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:164)
04-15 08:48:51.718 32430 32430 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:6944)
04-15 08:48:51.718 32430 32430 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
04-15 08:48:51.718 32430 32430 E AndroidRuntime: 	at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:327)
04-15 08:48:51.718 32430 32430 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1374)
04-15 08:48:51.718 32430 32430 E AndroidRuntime: Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void org.apache.cordova.CallbackContext.error(java.lang.String)' on a null object reference
04-15 08:48:51.718 32430 32430 E AndroidRuntime: 	at de.niklasmerz.cordova.fingerprint.Fingerprint.onCancelled(Fingerprint.java:420)
04-15 08:48:51.718 32430 32430 E AndroidRuntime: 	at de.niklasmerz.cordova.fingerprint.FingerprintAuthenticationDialogFragment.onCreateView(FingerprintAuthenticationDialogFragment.java:83)
04-15 08:48:51.718 32430 32430 E AndroidRuntime: 	at android.app.Fragment.performCreateView(Fragment.java:2611)
04-15 08:48:51.718 32430 32430 E AndroidRuntime: 	at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1276)
04-15 08:48:51.718 32430 32430 E AndroidRuntime: 	at android.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1549)
04-15 08:48:51.718 32430 32430 E AndroidRuntime: 	at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1611)
04-15 08:48:51.718 32430 32430 E AndroidRuntime: 	at android.app.FragmentManagerImpl.dispatchMoveToState(FragmentManager.java:3045)
04-15 08:48:51.718 32430 32430 E AndroidRuntime: 	at android.app.FragmentManagerImpl.dispatchActivityCreated(FragmentManager.java:2997)
04-15 08:48:51.718 32430 32430 E AndroidRuntime: 	at android.app.FragmentController.dispatchActivityCreated(FragmentController.java:179)
04-15 08:48:51.718 32430 32430 E AndroidRuntime: 	at android.app.Activity.performCreateCommon(Activity.java:7177)
04-15 08:48:51.718 32430 32430 E AndroidRuntime: 	at android.app.Activity.performCreate(Activity.java:7185)
04-15 08:48:51.718 32430 32430 E AndroidRuntime: 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1220)
04-15 08:48:51.718 32430 32430 E AndroidRuntime: 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2910)
04-15 08:48:51.718 32430 32430 E AndroidRuntime: 	... 9 more
`

On the other hand, if we don't invoke 'dismissAllowingStateLoss', the app crashes on 'onResume'. If we check the plugin state on 'onResume' as well, then a blank dialog is created because we return NULL on 'onCreateView'. So I decided to invoke 'dismissAllowingStateLoss' on 'onCreateView' to prevent showing a blank dialog and to prevent checking the plugin state on 'onResume' too. I think that the login is more clear like that.

This is the blank dialog created if we don't add 'dismissAllowingStateLoss'

![issue](https://user-images.githubusercontent.com/6747277/56114087-14aea600-5f60-11e9-9a5a-f7ac630b10f4.jpg)

Reply:
Thank you for the details. Looks like you found a strange bug there...

I am a little bit worried to just stop plugin execution. This could break developers code and lock apps up. Sounds weird but in this case if rather have apps crash and 'resume' properly from scratch.

Reply:
I understand your worry, but what would happen if the plugin isn't set up correctly like 'packageName' is NULL ? Would it work? That the plugin context is not set up as expected could be event more insecure than cancelling the biometric authentication process by dismissing the dialog. We could even check whether callbackContext is set to notify the app if you think that it's important, but I'm pretty sure that if 'packageName' is NULL 'callbackContext' will be NULL as well.

I'm committed to help you, tell me anything.

Reply:
I would love to understand first why the state is getting lost on resume? How can random variables be null just because the app was resumed? This could be Cordova bug. Do you have similar issues with other plugins? Can you image a test for another plugin to prove this and then maybe contact the Cordova devs? Do you have any other ideas why this is happening?

Your PR sounds like a good solution but I prefer to understand the root cause first if possible.

Reply:
Sorry for the delay but I'm involved in other stuff. Please, have a look at this documentation that is official for Android devs.

https://developer.android.com/guide/components/activities/activity-lifecycle#asem

Reply:
@NiklasMerz could you have a look at this?

Thanks in advance!

Reply:
Sorry for bothering you, but my team and I are waiting for the resolution of this PR. We don't want start using a fork of this plugin because this is being well-maintained.

Thanks in any case for your support and help.

Reply:
Sorry for the delayed reply. I am still not sure about this fix, because I was not able to reproduce this yet. Can you think of another way to check if the app is in an invalid state? I would like to understand the issue more before adding a workaround.

But I can understand your concerns about using a fork.

Reply:
No worries, I just wanted to know if you are still interested in get in touch with this. As for how to check if the app is in an invalid state, I don't know another way to check this in a Cordova plugin. Did you have time to have a look at the Google's official doc I sent you ?

https://developer.android.com/guide/components/activities/activity-lifecycle#asem

This doc explains somehow how to deal with this issue.